### ISSUE TITLE
Review the order of signed message and signed assertion

### ISSUE THREAD

**Issue by tngan** [tngan/samlify#219]
We continue the discussion here @rastko-florencehc. Since the code base is updated, so I will use the new one in this discussion.

https://github.com/tngan/samlify/blob/15db8b4cb2fc0cfadd074187adc8a11b14e7cc29/src/libsaml.ts#L341-L357

Here, the implementation assumes the IdP to sign message first then sign assertion, so when we verify it, we verify assertion signature first then message signature. At the moment we verify assertion signature, the saml content includes the message signature. Then we remove the signature node once it's verified.

However, I do think IdP should sign assertion first then sign the entire message to provide a full protection on data integrity.

**Comment by tngan** [comment]
@rastko-florencehc Confirmed there is a problem in the test case for our IdP to generate the signature, so I didn't realize there is an issue in signature verification for signing both assertion and entire message.

**Comment by tngan** [comment]
After tracing down the code for a while, the test case implementation has problem because of two reasons.

https://github.com/tngan/samlify/blob/b751ddc10cdc500d8102ba148975c7f65464ca28/test/flow.ts#L309-L323

#### 1. The callee `base64LoginResponse` from `createLoginResponse` signs message signature first then assertion signature, which should sign assertion signature first then entire message. However, these one is not the direct reason that cause the test case to be false positive.

https://github.com/tngan/samlify/blob/b751ddc10cdc500d8102ba148975c7f65464ca28/src/binding-post.ts#L133-L155

#### 2. Mutation of original xml document because of out-dated version of `xml-crypto`

https://github.com/yaronn/xml-crypto/blob/5c46e4f702a365fd9bb8124f21e0774b29908df5/lib/signed-xml.js#L792-L797

TLDR; In the current version of samlify, we are using the `xml-crypto` with out-dated version, so the line 796 above is not deep cloned in our current version. This will affect the signature construction in our IdP implementation, so there will be only one signature at all if the setting is configured to message and assertion signed. Since the original assertion signed document passed into `SignedXml.prototype.computeSignature` function will get mutated in `SignedXml.prototype. createSignedInfo -> createReferences  -> getCanonXml` function. 

https://github.com/yaronn/xml-crypto/blob/5c46e4f702a365fd9bb8124f21e0774b29908df5/lib/signed-xml.js#L647-L654

**Comment by tngan** [comment]
We will schedule another rc release within this week.

**Comment by rastko-florencehc** [comment]
@tngan Great! I'm glad this is resolved now. I'm only sorry I haven't seen this message yesterday I could've save you some debugging time since I found exact same thing!

**Comment by rastko-florencehc** [comment]
@tngan correct me if I'm wrong, but rc-6 now contains fixes for this problem?

**Comment by aagirusairam** [comment]
@tngan I'm looking for an unsigned SAML response from IDP. Currently, I encountered with 'ERR_ZERO_SIGNATURE'. Is Signing the SAML Response is required though we have configured the service provider to look for an unsigned response?

**Comment by tngan** [comment]
@aagirusairam Unsigned SAML response is not recommended in the SAML specification. See [4.1.3.5](http://docs.oasis-open.org/security/saml/v2.0/saml-profiles-2.0-os.pdf) in the specification..
